elks/arch/i86/kernel/irqtab.S

```diff
@@ -433,7 +217,8 @@ save_regs:
 	pop	(%si)		// DI
 	pop	2(%si)		// SI
 	pop	8(%si)		// DS
-	pop	%di		// Pointer to interrupt number
+	pop	%di		// Return offset is actually a pointer to the IRQ
+	pop	%ds		// Return segment of the dynamic handler = kernel DS
```

Comment:
I don't understand how the above `pop %ds` works. It isn't necessarily guaranteed that _irqit is only called by int_vector_set, which is passed a seperate 'kernel_ds' value anyways. Should pop %ds not be kernel DS, lines 223-226 will badly fail. Is this only being done to get DS for 244? Line 168 already sets DS = kernel_ds anyways. 

If Line 168 setting kernel DS is incorrect because we're executing in heap-allocated data segment, then Line 172 comparing _gint_count will fail also. So I don't get it.

Reply:
* Nothing else than a dynamic handler would call `_irqit`.
* Overwriting the same value in DS is a trick to avoid using another register, and for a minimal impact on a critical routine.

Reply:
The more I look at this, I realize new approach will not work with this code. DS needs to be properly set first thing in this routine or all code prior to your changes fails with improper DS, since CS now points to heap data. _irqit is a critical routine and can be called at any time. I have analyzed it heavily (https://github.com/jbruchon/elks/wiki/System-calls-and-interrupts) and it is time-proven to be not buggy. We need to be extremely careful rewriting it.

So the question really is: what benefit really is there to dynamic IRQ handling, since we're now at 3 extra instructions per interrupt, without solving problem. If problem is to reduce source clutter, perhaps just removing all CONFIG_IRQ_x options and always maintaining non-ifdefed table of `_irqtable` entry points in existing kernel code segment solves that problem. Then your routine can dynamically set handler within CS:_irqtable and point low-core vector to it . At this point, easier design, faster, and far less messy. That method still has its problems though, we lose configurability as to which low-memory entry points are patched, which needs discussion. If we keep _irqtable as CS entry point and patch low-memory at runtime, perhaps that's the best answer.

Reply:
> since CS now points to heap data

:interrobang:

Reply:
Look at the beginning of the routine!!! CS points to heap data:
```
_irqit:
 //
//	Make room
//
	push	%ds
	push	%si
	push	%di
//
//	Recover data segment
//
//	seg	cs
	mov	%cs:ds_kernel,%ds     <----- CS points to heap data, so this loads DS with garbage.
//
//	Determine which stack to use
//
	cmpw	$1,_gint_count   <---- this now compares a bogus word somewhere in memory
                              // kernel will now make incorrect decision on which stack to run on
                              // later incw _gintr_count at saveregs trashes memory from bad DS, etc
```
If DS was set correctly at the beginning of the routine, why would you need your additional `pop %ds`?
⁉️⁉️

Reply:
I get it now, sorry I missed that `int_vector_set` returns CS which is used to set the far call segment in the allocated trampoline. And the pop ds is required since far call was used to get to _irqit. Tricky.

Still uses 3 extra instructions per interrupt, forces code execution in data segment, and requires heap allocation which may fail. Using static _irqtable in code segment solves all those problems in better way, except that of compiled-in _irq_com1/2 fast serial driver.

Thus, to truly provide benefit, I suggest not using klugy `int_vector_set` secondary API call from serial driver routine, but instead use request_irq for fast serial drivers as well, but enhanced to take handler routine as far pointer with a flag to indicate whether or not to trampoline through _irqit or not. The third parameter to request_irq is currently unused and should be used as flag instead.

In the current kernel, there are only four calls to request_irq, so this PR is a lot of work and change without yet much benefit. If request_irq was required to be used from all drivers (not int_vector_set, etc), then option could be added that would allow both fast and slow serial drivers to be compiled in, for instance, and either one to be dynamically selected at runtime, even on ROM system. Using static _irqtable would not work in this case, resulting in a real benefit to this work. This would also provide your own stated goal of ensuring that drivers go through a standard process for gaining access to interrupts: use request_irq only. `enable_irq` could be removed in routines already calling request_irq.

Reply:
Here is another benefit : `ds_kernel` could now be dropped, as DS is already pushed by the CALLF in the dynamic trampoline. See commit 990a917. Still have to complete the work on the fast handlers as you told.

Reply:
Reworked the fast handlers in 1db6bc5.